### ISSUE TITLE
Using runtime built and dynamically-loaded dlls in Blazor Runtime causing issue in Kestrel hosting process.

### ISSUE THREAD

**Issue by StephenEhlers** [dotnet/runtime#76554]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

My goal is to Hot Reload a razor component dll reference that is not known to a Blazor WebAssembly App at compile time, but is built and referenced dynamically at runtime. I know that you can use a LazyAssemblyLoader to defer loading compile time known razor components to speed up application load times. But this is not my goal, I need to dynamically load a component that is not known to the website at compile time.

I have been able to get most of this to work, however it only correctly loads the changes in once before a strange hosting process bug appears and prevents subsequent updates from rendering until the process is terminated and restarted. There is no error in the dot NET hosting console application, but I believe the issue is inside of it because multiple browsers hitting the site will all manifest the issue, so its not a bug in the browser but in the hosting process.

This is the error in the browser console:
```
* Assertion at /__w/1/s/src/mono/mono/metadata/metadata.c:1096, condition `<disabled>' not met
      Uncaught (in promise) ExitStatus
```

### Expected Behavior

Since the first hot reload update succeeds, subsequent hot reload updates should also succeed and not result in inability to use the debug instance.

### Steps To Reproduce

**This first scenario is what is typically used for compile time known components, my scenario will make modifications to this so that you can more easily follow the logic of what I am doing**

1. Open Visual Studio and Create a new project.
2. Choose Blazor WebAssembly App template.
3. Choose, Framework .NET 6.0, None for authentication type, check Configure for HTTPS and ASP.NET Core Hosted.
4. Click the create button.
5. Test Building and Running the blazor application.
6. Close the debug window.
7. In visual studio add a new Razor Class library project to your solution.
8. Add project reference to your Razor Class project in the BlazorApp.Client project.
9. In the BlazorApp.Client project open the Pages folder and edit the Index.razor page.
10. Add a using statement for your razor class library
```razor
 @using RazorClassLibrary1;
```
11. Also add the default component from the Razor class library.
```
 <Component1></Component1>
```
12. Build and run the application.
13. See that your razor component now shows up on the page.
14. Test hot reloading by making changes to your component html and see it automatically refresh, that you can do this indefinitely without issues.

**Now the dynamic no project reference changes**

15. Remove project reference to Razor Class Library project from BlazorApp.Client project.

16. Go back to Index.razor and remove the Razor Class library using statement and replace the component reference with a DynamicComponent. Also add a button to execute the code to dynamically load the dll.
```razor
@if (_dynamicType != null)
{
    <DynamicComponent Type="_dynamicType" @ref="_dynamicControl"></DynamicComponent>
}
<div class="btn btn-primary" @onclick="handleLoadDynamic">Load Dynamic</div>
```

17. Remove using statement for Razor Component Library and instead add an inject call for HttpClient and IJSRuntime also add reference to System.Reflection.
```razor
@inject HttpClient Http;
@using System.Reflection;
@inject IJSRuntime JSRuntime;
```

18. Add an `@code` section to the page.
```razor
@code{
    private Type _dynamicType;
    private DynamicComponent _dynamicControl;

    private async void handleLoadDynamic()
    {
        string assemblyPath = $"dynamic/RazorClassLibrary1.dll"; // or whatever your dll is named
        string dynamicTypeName = $"RazorClassLibrary1.Component1"; // or whatever your components type name is

        try
        {
            byte[] dynamicBytes = await Http.GetByteArrayAsync(assemblyPath);
            var assembly = Assembly.Load(dynamicBytes);
            if (assembly != null)
            {
                _dynamicType = assembly.GetType(dynamicTypeName);
            }

            // IMPORTANT make sure state has changed or dynamic control will not be rendered.
            StateHasChanged();
        }
        catch(Exception ex)
        {
            await JSRuntime.InvokeVoidAsync("alert", $"Sorry, there doesn't seem to be a component called {dynamicTypeName}!");
            return;
        }
    }
}
```

19. Open the Razor Class Libraries properties tab and add a post-build even to copy its dll to the Blazor.Client projects wwwroot/dynamic folder.
```
(robocopy $(OutDir) $(ProjectDir)..\BlazorApp4\Client\wwwroot\dynamic\ RazorClassLibrary1.dll /it /NJH /NJS /NFL /NDL) ^& IF %ERRORLEVEL% GEQ 8 exit 1 
exit 0
```

20. Set the BlazorApp.Client project as the startup project and then run the debugger.
21. Click the Load Dynamic button and see the razor component load.
22. With debugger still running, make a change to the razor components UI, such as change the text in the component div and save it. You will notice the hot reload try to redraw the page, but the dll is dynamic and even clicking the Load Dynamic again nothing will change because the new dynamic dll is not in the website folder.
23. With debugger still running, execute an msbuild command on the Razor Class Library project.
	```
	 msbuild .\RazorClassLibrary1.csproj /targets:Build
	```
24. Now that the dynamic library was rebuilt and copied into our dynamic folder we can hit the Load Dynamic button again in our app.
25. Notice that the new dll loads successfully and our changes are present. (exactly what we want!)
26. Now repeat a change in the Razor Class Library, save it and re-run msbuild on it agian.
27. This time when you hit the Load Dynamic the changes do not render. 
28. Inspecting the console of the browser you will see this error message:
	```
	 * Assertion at /__w/1/s/src/mono/mono/metadata/metadata.c:1096, condition `<disabled>' not met
	 Uncaught (in promise) ExitStatus
	```
29. Try refreshing the browser, the error will keep throwing even before the Load Dynamic button is pressed.
30. Even stranger, if you open another browser (like firefox) and hit the same running debug site it will also result in the same error. Although it does give a little more detail than Chrome.
	```
	 WebSocket failed to connect. aspnetcore-browser-refresh.js:18:15
	 Firefox can’t establish a connection to the server at wss://localhost:44344/BlazorApp4.Client/.
	 * Assertion at /__w/1/s/src/mono/mono/metadata/metadata.c:1096, condition `<disabled>' not met
	```

So strangely, it seems that the error is actually stemming from the kestrel .net hosting app and not in the browser memory itself since it affects all browsers hitting the site.
However, the dotnet hosting console shows no errors and that everything is A-OK. 

### Exceptions (if any)

In Chrome:
```
* Assertion at /__w/1/s/src/mono/mono/metadata/metadata.c:1096, condition `<disabled>' not met
      Uncaught (in promise) ExitStatus
```
      
In FireFox:
WebSocket failed to connect. aspnetcore-browser-refresh.js:18:15
```
Firefox can’t establish a connection to the server at wss://localhost:44344/BlazorApp4.Client/.
	* Assertion at /__w/1/s/src/mono/mono/metadata/metadata.c:1096, condition `<disabled>' not met
```

### .NET Version

6.0

### Anything else?

I found that some others were getting the same top level error as this here:
https://github.com/dotnet/runtime/issues/42628

However it seems that they were getting to the exception with an entirely different use case than mine.
[BlazorApp4.zip](https://github.com/dotnet/aspnetcore/files/9630507/BlazorApp4.zip)

**Comment by mkArtakMSFT** [comment]
Discussed this with @lewing. The error you're seeing is coming from the runtime. @lewing will pull in the right folks to understand what's actually happening and see how we will move forward from there.

Thanks @lewing !

**Comment by LunicLynx** [comment]
From what i understand is that hot reloading doesn't do what you expect from it. As far as i know it is incremental, meaning everything that has happened will happen again for the session since you started the server.
So reloading or changing the browser doesn't really reset the session. Only solution is to restart.

**Comment by lambdageek** [comment]
If I understand the workflow correctly I think this isn't a hot reload issue, but rather it's two things:

1. I think that perhaps for some reason Kestrel is serving a subset of the updated assembly (ie if after the update the razor component library .dll grew from X to X+Y bytes, it's only serving the first X bytes)
2. the runtime when doing `Assembly.Load(byte[])` succeeds, but subsequently either `assembly.GetType(dynamicTypeName)` or actually making use of `_dynamicType` needs something past the end of the loaded bytes which ends up asserting (in this case in `mono_metadata_string_heap`, but in separate testing I also got exceptions such as `System.BadImageFormatException: Method has zero rva` when calling a method from the loaded type).

I will need to investigate more to find out exactly what is happening...

**Comment by StephenEhlers** [comment]
@lambdageek yes, as one of my follow up posts on this thread adds, the same workflow and code with IIS express works without issue. So I agree, with you it is probably something in Kestrel that is not fully reinitialize properly when the same file is served up after modification more than once.

**Comment by lewing** [comment]
@thaystg do you have any idea here?

**Comment by thaystg** [comment]
Hello, @StephenEhlers, I was trying to reproduce the issue, but for me it's working, I can change the code in the RazorClassLibrary1, rebuild it using dotnet build, and the click on the button and I don't see any exception, I did this process more than 5 times, and no assertion.
I would like to check with you which version of .net6 are you using?

**Comment by StephenEhlers** [comment]
Hi @thaystg, I downloaded the project I originally uploaded above and tried to reproduce again however I can no longer get the error. The project is still referencing .net 6 as it originally did but my build tools from command line are all using 7.0. Either way using Kestrel as the debug target now seems to work fine, I did updates about 10 or so and it was solid. 

I am using this dynamic loading process in an ongoing project, so I will switch back from IISExpress to Kestrel and reopen this case if i see a re-occurrence. For now I see no reason to leave this issue open.

Thanks for following up.
